c-dependencies/js-compute-runtime/builtins/backend.cpp

```diff
@@ -1,22 +1,602 @@
+   */
+  std::vector<Cipher> defaultSort(std::vector<Cipher> &ciphers) {
+    std::vector<Cipher> result;
+    result.reserve(ciphers.size());
+    std::vector<Cipher> ecdh;
+
+    /* Everything else being equal, prefer ephemeral ECDH over other key exchange mechanisms */
+    std::vector<Cipher> eecdh;
+    std::copy_if(ciphers.begin(), ciphers.end(), std::back_inserter(result),
+                 byKeyExchange(KeyExchange::EECDH));
+
+    /* AES is our preferred symmetric cipher */
+    auto aes = {Encryption::AES128, Encryption::AES128GCM, Encryption::AES256,
+                Encryption::AES256GCM};
+
+    /* Now arrange all ciphers by preference: */
+    std::copy_if(ecdh.begin(), ecdh.end(), std::back_inserter(result), byEncryption(aes));
+    std::copy_if(ciphers.begin(), ciphers.end(), std::back_inserter(result), byEncryption(aes));
+
+    /* Add everything else */
+    result.insert(result.end(), ecdh.begin(), ecdh.end());
+    result.insert(result.end(), ciphers.begin(), ciphers.end());
+
+    /* Move ciphers without forward secrecy to the end */
+    std::vector<Cipher> ciphersWithoutForwardSecrecy;
+    std::copy_if(result.begin(), result.end(), std::back_inserter(ciphersWithoutForwardSecrecy),
+                 byKeyExchange(KeyExchange::RSA));
+    moveToEnd(result, ciphersWithoutForwardSecrecy);
+
+    strengthSort(result);
+    return result;
+  }
+
+  auto byProtocol(Protocol val) -> std::function<bool(const Cipher &)> {
+    return [val](auto &c) { return c.protocol == val; };
+  }
+
+  auto byKeyExchange(KeyExchange val) -> std::function<bool(const Cipher &)> {
+    return [val](auto &c) { return c.kx == val; };
+  }
+
+  auto byAuthentication(Authentication val) -> std::function<bool(const Cipher &)> {
+    return [val](auto &c) { return c.au == val; };
+  }
+
+  auto byEncryption(std::set<Encryption> vals) -> std::function<bool(const Cipher &)> {
+    return [vals](auto &c) { return vals.find(c.enc) != vals.end(); };
+  }
+
+  auto byEncryption(Encryption val) -> std::function<bool(const Cipher &)> {
+    return [val](auto &c) { return c.enc == val; };
+  }
+
+  auto byEncryptionLevel(EncryptionLevel val) -> std::function<bool(const Cipher &)> {
+    return [val](auto &c) { return c.level == val; };
+  }
+
+  auto byMessageDigest(MessageDigest val) -> std::function<bool(const Cipher &)> {
+    return [val](auto &c) { return c.mac == val; };
+  }
+
+  std::vector<std::string> split(std::string s, std::string_view delimiter) {
+    size_t pos_start = 0, pos_end, delim_len = delimiter.length();
+    std::string token;
+    std::vector<std::string> res;
+
+    while ((pos_end = s.find(delimiter, pos_start)) != std::string::npos) {
+      token = s.substr(pos_start, pos_end - pos_start);
+      pos_start = pos_end + delim_len;
+      res.push_back(token);
+    }
+
+    res.push_back(s.substr(pos_start));
+    return res;
+  }
+
+  std::vector<std::string> splitCipherSuiteString(std::string string) {
```

Comment:
Looking at how the result is used lower down, I think the input here could be a `std::string_view`, returning a vector of string_views.
```suggestion
  std::vector<std::string_view> splitCipherSuiteString(std::string_view string) {
```